components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.tenant.initializer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/tenant/initializer/listener/ServerStartupListener.java

```diff
@@ -46,29 +61,84 @@
 
     @Override
     public void completedServerStartup() {
-        ScheduledThreadPoolExecutorImpl.waitForAdminServiceActivation();
+        ScheduledThreadPoolExecutorImpl.waitAndInitialize();
     }
 
     /**
-     * Inner class for holding API Synchronization until Authentication Admin Service starts
-     * NOTE: The time delay introduced here is a temporary fix.
-     * This is tracked by issue https://github.com/wso2/cloud/issues/1581 and will be handled shortly
+     * Inner class for holding API and Throttling Synchronization until Authentication Admin Service starts
      */
-    static class ScheduledThreadPoolExecutorImpl {
-        static void waitForAdminServiceActivation() {
-            ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
-            executor.schedule(new Runnable() {
-                public void run() {
+    static class ScheduledThreadPoolExecutorImpl implements Runnable {
+        private int retryCount = 4;
+        private int executionCount = 0;
+        private static ScheduledThreadPoolExecutor executor;
+        private static String adminName;
+        private static char[] adminPwd;
+        private static String url;
+
+        public static void waitAndInitialize() {
+            try {
+                String mgtTransport = CarbonUtils.getManagementTransport();
+                AxisConfiguration axisConfiguration = ServiceReferenceHolder
+                        .getContextService().getServerConfigContext().getAxisConfiguration();
+                int mgtTransportPort = CarbonUtils.getTransportProxyPort(axisConfiguration, mgtTransport);
+                if (mgtTransportPort <= 0) {
+                    mgtTransportPort = CarbonUtils.getTransportPort(axisConfiguration, mgtTransport);
+                }
+                url = mgtTransport + "://localhost:" + mgtTransportPort + "/services/";
```

Comment:
is it ok to hard cord the host here?

Reply:
Actually I tried to get it using org.wso2.carbon.utils.NetworkUtils but it returns the IP address. That results in an SSL issue and even setting truststore and keystore properties in code level didn't fix that. Is there another way of handling this?

Reply:
Did you try using the CarbonUtils.getServerURL() method?

Reply:
That returns local://services/ as the URL.

Reply:
@amandaJayanetti - I think you can get the hostname using the following code

CarbonUtils.getServerConfiguration().getFirstProperty("HostName");

Please add the "HostName" as a constant

Reply:
That only works if you have set the hostname in carbon.xml otherwise it returns null

Reply:
Can you try CarbonUtils.getServerURL() again? This will return local://services/ .Try to execute your code with this and see..